### ISSUE TITLE
Batch cursors should have a lifetime

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/differential-dataflow#66]
The cursors exposed by batches provide references to the underlying keys and values, but they are only valid for the borrow of the cursor associated with the call. This means that users of these cursors cannot extract a collection of references and work on them, even though in all current cursors this would be fine (manipulating the cursor does not mutate the underlying data).

I believe the right way to do this is to extend the `Cursor` trait to have a lifetime `'batch`, corresponding to the lifetime of the underlying batch:

```rust
pub trait Cursor<'batch, K, V, T, R> {
    // ...
```

Ideally this lifetime would get captured when we construct the cursor using `BatchReader::cursor()`, like so:

```rust
pub trait BatchReader<K, V, T, R> {

	/// The type used to enumerate the batch's contents.
	type Cursor: Cursor<'batch, K, V, T, R>;
	/// Acquires a cursor to the batch's contents.
	fn cursor(&'batch self) -> Self::Cursor;
```

but this doesn't work, because the `'batch` lifetime is part of the `Self::Cursor` type, and for that to work I think we need to have the `'batch` lifetime as part of the `BatchReader` trait. And, one expects, this lifetime taints just about everything upward. Eventually we are able to write `X: for<'batch> Trait<'batch, ...>`, but I haven't figured out where this happens. 

Perhaps just in `TraceReader`; it could be that it is as simple as

```rust
pub trait TraceReader<Key, Val, Time, R> {

	/// The type of an immutable collection of updates.
	type Batch: for<'batch> BatchReader<'batch, Key, Val, Time, R>+Clone+'static;
```

but I think these constraints would still need to be re-iterated at each point that an implementor of `TraceReader` is bound. Because they are requirements Rust imposes but not uses, for now.

Anyhow, doing all of this, or something similar, would result in the ability for the cursor to return references with lifetime `'batch` which would allow a consumer to stash and work with multiple references, rather than cloning everything.

**Comment by frankmcsherry** [comment]
Some thoughts after attempting to start this, and then stopping:

It seems like we'll want at least two types as part of a cursor:

1. The first is the owned "proof of life" a borrow against which grants a lifetime that then gets to show up in the references to keys and values and such. Everywhere we currently have an owned cursor, we will probably instead want to hold on to this owned proof of life (I was calling it a "capture", but that code went away).

2. The second is the cursor, which borrows the capture using the lifetime that shows up in the references. The cursor can now zoom around and change its state and all that, while remaining clear that everything with the outer lifetime remains stable. This freezes the capture for the duration, which is fine because it is immutable anyhow.

This is all well and good, but it also seems like we will need to pack up the cursor at some point so that we can remember how far we got. That is, if we navigated around for a bit and then, without finishing, decided to stop for the moment, we would need to be able to record something about our progress that would allow us to rehydrate the cursor when we wanted to use it again. Previously, the owned cursor just did that for us and there were no problems. Since the cursor (in this setting) has a borrow in it, we can't just stash it next to the capture because lifetimes. There are probably lots of horrible ways to do this (an associated type `State` that we can extract from / seed the cursor?).

Perhaps a "more clear" (???) version is to have the immutable capture and some mutable state, and the cursor then grabs an immutable borrow of the capture and a mutable borrow of the state. The cursor is now relatively little other than the code that screws around with the state and looks in the captured data.

**Comment by frankmcsherry** [comment]
Perhaps it is simpler than all this. 

Imagine a hypothetical design where the "proof of life" provides the navigation, but using an immutable borrow of itself and updating an external state:

```rust
trait Navigable<K, V> {
    type CursorState;
    seek<'a>(&'a self, state: &mut Self::CursorState, key: &K) -> Option<&'a V>;
}
```

The intent is that we take a borrow of the navigable data outside of whatever loops we have in mind, within which we want to stash lots of references to keys and values and such, mint up a new `mut cursor_state`, and use that to wander around in the data.

This could impose some restrictions on how we navigate around in data, because it seems simpler than the HRL approach above; perhaps it will be hard to add new modes of navigation for existing piles of data, because the data source needs to implement them, whereas the approach above lets the cursor define the implementation without wrapping the proof of life in some new type... (?)

**Comment by frankmcsherry** [comment]
Ok, wow. Who ever thought it would be so hard to produce a `&'a ()`. Will need to introduce an `EmptyLayer` just to hold on to a `()` so that I can return a long-lived reference to it.

Thank you, Rust.

**Comment by frankmcsherry** [comment]
This totally got implemented in #72. I guess I forgot to close it!
